cpp/src/arrow/compute/kernel.h

```diff
@@ -555,8 +505,7 @@ struct Kernel {
 /// endeavor to write into pre-allocated memory if they are able, though for
 /// some kernels (e.g. in cases when a builder like StringBuilder) must be
 /// employed this may not be possible.
-using ArrayKernelExec =
-    std::function<Status(KernelContext*, const ExecSpan&, ExecResult*)>;
+typedef Status (*ArrayKernelExec)(KernelContext*, const ExecSpan&, ExecResult*);
```

Comment:
Using plain function pointers makes debugging this code much easier (`std::function` adds many levels to gdb stack traces); we should use function pointers for the other kernel APIs in this file, too